src/modules/tokens.ts

```diff
@@ -1,35 +1,231 @@
-import { Address, BigInt } from "@graphprotocol/graph-ts"
+import { Address, BigDecimal, BigInt, dataSource } from "@graphprotocol/graph-ts"
 import { ERC20 } from "../../generated/GPV2Settlement/ERC20"
-import { Token } from "../../generated/schema"
+import { Token, TokenDailyTotal, TokenHourlyTotal, TokenTradingEvent } from "../../generated/schema"
+import { ZERO_BD, ZERO_BI, MINUS_ONE_BD, ONE_BI } from "../utils/constants"
+import { getPrices } from "../utils/getPrices"
+import { totals } from "./totals"
+import { getDayTotalTimestamp, getHourTotalTimestamp } from "../utils/timeframeTimestamp";
 
 const DEFAULT_DECIMALS = 18
 
 export namespace tokens {
 
-    export function getOrCreateToken(tokenAddress: Address, timestamp: BigInt): Token {
-        let tokenId = tokenAddress.toHexString()
-        let token = Token.load(tokenId)
-      
-        if (!token) {
-          token = new Token(tokenId)
-          token.address = tokenAddress
-          token.firstTradeTimestamp = timestamp
-      
-          let erc20Token = ERC20.bind(tokenAddress)
-          let tokenDecimals = erc20Token.try_decimals()
-          let tokenName = erc20Token.try_name()
-          let tokenSymbol = erc20Token.try_symbol()
-          token.decimals = !tokenDecimals.reverted
-              ? tokenDecimals.value
-              : DEFAULT_DECIMALS
-          token.name = !tokenName.reverted ? tokenName.value : ""
-          token.symbol = !tokenSymbol.reverted ? tokenSymbol.value : ""
-          token.totalVolume = BigInt.fromI32(0)
-
-          token.save()
+  export function getOrCreateToken(tokenAddress: Address, timestamp: BigInt): Token {
+    let tokenId = tokenAddress.toHexString()
+    let token = Token.load(tokenId)
+    let network = dataSource.network()
+
+    if (!token) {
+      token = new Token(tokenId)
+      token.address = tokenAddress
+      token.firstTradeTimestamp = timestamp
+
+      let erc20Token = ERC20.bind(tokenAddress)
+      let tokenDecimals = erc20Token.try_decimals()
+      let tokenName = erc20Token.try_name()
+      let tokenSymbol = erc20Token.try_symbol()
+      token.decimals = !tokenDecimals.reverted
+        ? tokenDecimals.value
+        : DEFAULT_DECIMALS
+      token.name = !tokenName.reverted ? tokenName.value : ""
+      token.symbol = !tokenSymbol.reverted ? tokenSymbol.value : ""
+      token.totalVolume = ZERO_BI
+      if (network == 'xdai') {
+        let tokenPrices = getPrices(tokenAddress)
+        if (tokenPrices.get("usd") != MINUS_ONE_BD &&
+          tokenPrices.get("eth") != MINUS_ONE_BD) {
+          token.priceUsd = tokenPrices.get("usd")
+          token.priceEth = tokenPrices.get("eth")
+        }
+      } else {
+        token.priceEth = ZERO_BD
+        token.priceUsd = ZERO_BD
```

Comment:
Curiosity: why can xdai have token prices while mainnet/rinkeby starts with 0?

Reply:
I'm not sure if I understood the question, but I added zero as a default on all tokens due to zero is neutral for addition.

Reply:
I mean, I understood that only gchain prices were initiated with `0` given the conditional